### ISSUE TITLE
`watch`/`toFlow` doesn't emit empty-cache response with nulled our `data`, it never emits/fails instead

### ISSUE THREAD

**Issue by mateuszkwiecinski** [apollographql/apollo-kotlin#3976]
**Summary**
I wasn't sure if that's a bug or an expected behaviour change. But looking at the docs, it feels more like an unwanted side effect 👀
When the cache is empty 2.x emitted new `ApolloResponse` with null `data`. `3.2.0` either throws or waits until it sees valid cache state
<img width="226" alt="image" src="https://user-images.githubusercontent.com/36954793/160824416-cbd63e73-e8f5-430a-9e61-f402a1227fac.png">

**Version**
`3.2.0`

**Description**
I could probably leverage parameters added in https://github.com/apollographql/apollo-kotlin/issues/3786 to write a custom extension that emits _empty_ `ApolloResponse` when `CacheMissException` gets thrown, but I thought I should ask here first.
The minimal repro steps would be something like:
```kotlin
val response = apollo.query(SomeQuery())
	.fetchPolicy(FetchPolicy.CacheOnly)
    .watch()
    .first()
   
check(response.data == null)
```

I'll also describe our use case: our queries either continuously watch the cache without making network calls or make a single network call to pre-fetch/refresh the data. We basically display cache state on the UI and refresh the state when _needed_, apollo cache is considered as a source of the truth. 
Previously, we've been able to distinguish 3 cache states: 
1. Unknown cache state - the moment until first emission is received
2. Empty cache state - when we've received valid `ApolloResponse` with `null` `data` field
3. Non-empty cache state - when `ApolloResponse` with filled `data` field
(we were also ready to handle fourth: internal apollo error where the whole Flow throws, but that's unrelated and it never happened so far 👌)

With current 3.x behaviour, it doesn't seem possible to easily distinguish 1) and 2) 👀 I browsed documentation and release notes and I couldn't find the answer to my question. Seeing the `ApolloResponse#data` is still nullable it doesn't feel that much _off_ to have it null on empty cache 🤔 

Am I wrong to expect `null` `data` to be emitted? And if I am: what's the _recommended_ way to know when the cache is considered "empty"? 

Full repro: https://github.com/mateuszkwiecinski/apollo-playground/commit/62ffffed32c56e3df2d33a1486897ba918c8c7e7

**Comment by martinbonnin** [comment]
Thanks as always for the very clear issues 💜 . This one is a hard one. This was an intentional change. The reasons are:

1. Having `response.data == null` is also a valid network response. It's an edge case but it's still valid for an example if an error happens on a non-nullable root field and the error has to "bubble" up. This means it'd be impossible to distinguish between this case and the cache miss case.
2. Throwing a `CacheMissException` instead of emitting `data == null` gives more structured information about the cache misses like the missing cache keys that can be used for troubleshooting or logging.

Also (but this would have to be confirmed), it _felt_ like most use cases didn't need the cache miss information and having to handle the extra `null` would cause a bunch of extra logic at the call site.

It's true that it doesn't work for the use case you're describing though. If you need that back, I'd recommend doing something like this:

```kotlin
apolloClient.query(query)
            .fetchPolicy(FetchPolicy.CacheOnly)
            .refetchPolicy(FetchPolicy.CacheOnly)
            .watch(fetchThrows = true, refetchThrows = true)
            .catch {
              if (it is CacheMissException) {
                emit(null)
              } else {
                throw it
              }
            }
```

Agreed it feels a bit off. In the long run, we might want to fine tune these APIs. Maybe emit sealed classes or something like this so we have all the items? Would that help?

**Comment by lwasyl** [comment]
> Having response.data == null is also a valid network response.

Are error responses saved in the cache? There was a flag for it a while back, but if it's been removed for 3.x, then maybe cache doesn't need to concern itself with this case?

As for the proposed workaround – I'm not sure it will work, doesn't throwing an exception in the flow cancel it? Which means a flow that once throws `CacheMissException` will never emit again? And restarting the collection right away doesn't make much sense as it'd immediately throw again. We'd need to wait until the cache populated, but there's no signal for that, and we're back to basically the original problem that there's no flow emitting cache state 😬

**Comment by martinbonnin** [comment]
Trying to make sure I understand the use case:

>  We basically display cache state on the UI and refresh the state when needed, apollo cache is considered as a source of the truth.
> 1. Unknown cache state - the moment until first emission is received

Is this when you display a loading state?

> 2. Empty cache state - when we've received valid ApolloResponse with null data field

Is this where you trigger the network request?  

> Non-empty cache state - when ApolloResponse with filled data field

I guess this is the case where the data is displayed.

Am I correct? If yes, any reason you're not using a `CacheFirst` refetchPolicy in 2. And leave the flow update its own data?

**Comment by mateuszkwiecinski** [comment]
>Is this when you display a loading state?

Looking at the code it seems like we don't do anything specific in that state. An empty UI waits for first emission. The loading state is a side effect of a network call (which happens separately, and isn't necessarily the result of the screen being opened). 

>Is this where you trigger the network request?

Depending on the use case. We either show an "empty" state, `flatMap` to some other input source as a default value provider, and sometimes, independently, we trigger a network request basing on the cache value being null or not. 

>I guess this is the case where the data is displayed.

That's correct. In almost all cases, it gets transformed and combined with other input sources, but in general yes, we enter the "happy path" here and present loaded data to the user

> any reason you're not using a CacheFirst refetchPolicy in 2. And leave the flow update its own data?

I'd answer that it's because of our _architecture™_, all the decisions that were made over last 2-3 years. 
We couldn't reliably use `CacheFirst` in 2.x as it's been cancelling the whole flow if the network call failed, resulting in no further updates being emitted. AFAIU this behaviour got significantly improved in 3.x and `watch()` re-subscribes to cache updates after initial failure, so to overcome the original issue, we've built an additional layer over Apollo to leverage the normalized cache and not rely on previous cache-strategies too much. We enforced unidirectional flow, where observing the cache can't ever fail or have any side effects (like network calls).
We're quite satisfied with the architecture we came up with (it's havily inspired by [CQS](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation)), as it addressed most of the challenges we were facing. For example, using `CacheFirst`, in multiple places, called at once, resulted in multiple network calls being made which led to flaky UI tests (there was a race condition, where sometimes the second `CacheFirst` was called after the first query finished writing data to cache, and sometimes it didn't).


☝️ Hope that's helpful to understand our situation/point of view 😃

**Comment by martinbonnin** [comment]
Thanks for the context! A new `emitCacheMisses` API is available with https://github.com/apollographql/apollo-kotlin/pull/3980:

```kotlin
      apolloClient.query(query)
          .fetchPolicy(FetchPolicy.CacheOnly)
          .emitCacheMisses(true)
          .watch()
          .collect {
            channel.send(it.data)
          }
```

It's simple enough that it shouldn't be too hard to maintain but we might want to streamline the ergonomics moving forward. I think the tension here is the one of `suspend` vs `Flow`:

1. `suspend` cases like `fetchPolicy(CacheFirst).execute()` want a single response value
2. `Flow` cases like `fetchPolicy(CacheOnly).watch().collect {}` may or may not want the cache misses emitted as regular items.

Given this is opt-in, I think it's fine for now, let me know what you think.

Back to your use case, one thing I'm still not 100% sure about is what triggers the network:

> We enforced unidirectional flow, where observing the cache can't ever fail or have any side effects (like network calls).

I get that but there still needs something that does the network call, right? If you're doing a network call every time you have a cache miss, you're back to square one where multiple concurrent watchers may emit multiple cache misses and you end up with concurrent network requests fetching the same data. Or do you serialize the network requests somehow?

**Comment by mateuszkwiecinski** [comment]
The `emitCacheMisses` looks great 👌 Thank you! 

> but there still needs something that does the network call, right?

There is usually a place which triggers a network call, but it isn't necessarily bound with cache being empty. Empty cache could be considered a _valid_ state, where an "empty-state" UI gets presented to the user. The refresh can happen when the app receives a push message or some other external, unrelated to the _current_ screen event. A scenario where network isn't ever reached could also take place.

> you're back to square one where multiple concurrent watchers may emit multiple cache misses and you end up with concurrent network requests fetching the same data.

In most cases, we have precise control over which resource performs a network call, so the app can block/hide/ignore a subset of actions which would lead to the same resource being refreshed (example: if a network call is triggered in the background by a push notification, doing a pull-to-refresh action on the UI will piggyback on existing refresh call, the app will make single network call). 
Although, you're partially correct here 😛 We do still have edge cases where we make superfluous network calls, but that's something we plan to tackle soon™ (we've briefly discussed a mechanism similar to query batching `apollo-kotlin` already provides)

And just to be clear: most probably we _could_ rely on Apollo defaults and use predefined Policies everywhere (especially with all the 3.x improvements 🚀 ), but we tried not to _leak_ current apollo implementation to "our" code. At the same time I don't believe we're doing "crazy" stuff, unidirectional data flow doesn't sound that uncommon 😛

**Comment by BoD** [comment]
`emitCacheMisses` is now available in the just released v3.2.1.
